code/game/objects/items/weapons/grenades/decompiler.dm

```diff
@@ -0,0 +1,142 @@
+/obj/item/grenade/decompiler
+	desc = "It is set to detonate in 5 seconds. It will create an unstable singularity that will break nearby objects down into purified matter cubes."
+	name = "decompiler grenade"
+	icon = 'icons/obj/grenade.dmi'
+	icon_state = "delivery"
+	item_state = "flashbang"
+	origin_tech = "{'materials':3,'magnets':2,'exoticmatter':3}"
+	matter = list(
+		MAT_PHORON = MATTER_AMOUNT_REINFORCEMENT,
+		MAT_SUPERMATTER = MATTER_AMOUNT_TRACE
+	)
+
+/obj/item/grenade/decompiler/detonate()
+	var/spawn_loc = get_turf(src)
+	qdel(src)
+	new /obj/effect/decompiler(spawn_loc)
+
+/obj/effect/decompiler
+	name = "unstable singularity"
+	desc = "Wow, this really sucks."
+	icon = 'icons/effects/96x96.dmi'
+	icon_state = "singularity_s3"
+	pixel_w = -32
+	pixel_z = -32
+	alpha = 0
+
+	var/pull_range = 7
+	var/eat_range = 1
+	var/lifetime = 10 SECONDS
+	var/expiry_time
+	var/list/decompiled_matter
+	
+/obj/effect/decompiler/Initialize()
+	. = ..()
+	expiry_time = world.time + lifetime
+	START_PROCESSING(SSobj, src)
+	decompiled_matter = list()
+	fade_in()
+
+/obj/effect/decompiler/proc/fade_in()
+	visible_message(SPAN_DANGER("\A [src] forms, reaching out hungrily!"))
+	playsound(loc, 'sound/magic/ethereal_enter.ogg', 75, FALSE)
+	set_light(0.8, 0, 4.5, l_color = LIGHT_COLOR_PURPLE)
+	var/matrix/M = matrix()
+	M.Scale(0.01)
+	transform = M
+	animate(src, transform = null, alpha = 255, easing = SINE_EASING, time = 5)
+
+/obj/effect/decompiler/proc/fade_out()
+	set waitfor = FALSE
+	playsound(loc, 'sound/magic/ethereal_exit.ogg', 75, FALSE)
+	var/matrix/M = matrix()
+	M.Scale(0.01)
+	animate(src, alpha = 0, transform = M, time = 5)
+	sleep(5)
+	var/dump_cubes = get_turf(src)
+	if(dump_cubes)
+		visible_message(SPAN_DANGER("\The [src] collapses!"))
+		for(var/mat in decompiled_matter)
+			var/sheet_amount = Floor(decompiled_matter[mat]/SHEET_MATERIAL_AMOUNT)
+			if(sheet_amount > 0)
+				while(sheet_amount > 100)
+					var/obj/item/stack/material/cubes/cubes = new (dump_cubes, 100, mat)
+					cubes.pixel_w = rand(-8, 8)
+					cubes.pixel_z = rand(-8, 8)
+					sheet_amount -= 100
+				var/obj/item/stack/material/cubes/cubes = new(dump_cubes, sheet_amount, mat)
+				cubes.pixel_w = rand(-8, 8)
+				cubes.pixel_z = rand(-8, 8)
+	qdel(src)
+
+/obj/effect/decompiler/Process()
+
+	FOR_DVIEW(var/atom/movable/thing, pull_range, loc, INVISIBILITY_LEVEL_TWO)
+		if(thing.simulated && !thing.anchored)
+			step_towards(thing, src)
+	END_FOR_DVIEW
+
+	if(world.time > expiry_time)
+		fade_out()
+		return
```

Comment:
Feel like you should return PROCESS_KILL, as this proc is waitfor false and sleeps before qdeling, so this may fire multiple times.